segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,380 @@
+        }
+
+        try (InputStream reader = client.readObjectStream(config.getBucket(),
+                config.getRoot() + handle.getSegmentName(), Range.fromOffsetLength(offset, length))) {
+
+            if (reader == null) {
+                throw new StreamSegmentNotExistsException(handle.getSegmentName());
+            }
+
+            int bytesRead = StreamHelpers.readAll(reader, buffer, bufferOffset, length);
+
+            LoggerHelpers.traceLeave(log, "read", traceId, bytesRead);
+            return length;
+        }
+    }
+
+    private StreamSegmentInformation syncGetStreamSegmentInfo(String streamSegmentName) {
+        long traceId = LoggerHelpers.traceEnter(log, "getStreamSegmentInfo", streamSegmentName);
+        S3ObjectMetadata result = client.getObjectMetadata(config.getBucket(),
+                config.getRoot() + streamSegmentName);
+
+        AccessControlList acls = client.getObjectAcl(config.getBucket(), config.getRoot() + streamSegmentName);
+
+        boolean canWrite = acls.getGrants().stream().anyMatch((grant) -> grant.getPermission().compareTo(Permission.WRITE) >= 0);
+
+        StreamSegmentInformation information = new StreamSegmentInformation(streamSegmentName,
+                result.getContentLength(), !canWrite, false,
+                new ImmutableDate(result.getLastModified().toInstant().toEpochMilli()));
+        LoggerHelpers.traceLeave(log, "getStreamSegmentInfo", traceId, streamSegmentName);
+        return information;
+    }
+
+    private boolean syncExists(String streamSegmentName) {
+        ListObjectsResult result = null;
+        try {
+            result = client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName);
+            return !result.getObjects().isEmpty();
+        } catch (S3Exception e) {
+            if ( e.getErrorCode().equals("NoSuchKey")) {
+                return false;
+            } else {
+                throw e;
+            }
+        }
+    }
+
+    private SegmentProperties syncCreate(String streamSegmentName) throws StreamSegmentExistsException {
+        long traceId = LoggerHelpers.traceEnter(log, "create", streamSegmentName);
+
+        if (!client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName).getObjects().isEmpty()) {
+            throw new StreamSegmentExistsException(streamSegmentName);
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getBucket(),
+                config.getRoot() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
```

Comment:
There is no deadlock as there is no onwership in extended S3 (and filesystemstorage) implemented. It depends on same sequence of operations being carried out in the same order. That is what is *so Idempotent* about them. 
```
Each block of data has an initial offset assigned to it. The data and the initial offset is stored in DurableLog.
In case of retries, Pravega always writes the same data to the same offset. * As a result the only flow when a write call is made to the same offset twice is when ownership of the segment changes from one host to another and both the hosts are writing to it.
```